server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java

```diff
@@ -0,0 +1,261 @@
+      throttler.incrementAssignedReplicas(tier);
+
+      final TierLoadingState replicatingInTier = currentlyReplicatingSegments
+          .computeIfAbsent(tier, t -> new TierLoadingState(throttler.getMaxLifetime()));
+      replicatingInTier.addSegment(segment.getId(), server.getServer().getHost());
+      callback = success -> replicatingInTier.removeSegment(segment.getId());
+    } else {
+      throttler.incrementThrottledReplicas(tier);
+      return false;
+    }
+
+    try {
+      if (!server.startOperation(segment, SegmentState.LOADING)) {
+        return false;
+      }
+
+      server.getPeon().loadSegment(
+          segment,
+          isPrimary ? SegmentAction.LOAD_AS_PRIMARY : SegmentAction.LOAD_AS_REPLICA,
+          callback
+      );
+      return true;
+    }
+    catch (Exception e) {
+      server.cancelOperation(segment, SegmentState.LOADING);
+      return false;
+    }
+  }
+
+  public boolean dropSegment(DataSegment segment, ServerHolder server)
+  {
+    try {
+      if (!server.startOperation(segment, SegmentState.DROPPING)) {
+        return false;
+      }
+
+      server.getPeon().dropSegment(segment, null);
+      return true;
+    }
+    catch (Exception e) {
+      server.cancelOperation(segment, SegmentState.DROPPING);
+      return false;
+    }
+  }
+
+  public boolean moveSegment(
+      DataSegment segment,
+      ServerHolder fromServer,
+      ServerHolder toServer,
+      int maxLifetimeInBalancingQueue
+  )
+  {
+    final TierLoadingState segmentsMovingInTier = currentlyMovingSegments.computeIfAbsent(
+        toServer.getServer().getTier(),
+        t -> new TierLoadingState(maxLifetimeInBalancingQueue)
+    );
+    final LoadQueuePeon fromServerPeon = fromServer.getPeon();
+    final LoadPeonCallback moveFinishCallback = success -> {
+      fromServerPeon.unmarkSegmentToDrop(segment);
+      segmentsMovingInTier.removeSegment(segment.getId());
+    };
+
+    // mark segment to drop before it is actually loaded on server
+    // to be able to account for this information in BalancerStrategy immediately
+    toServer.startOperation(segment, SegmentState.MOVING_TO);
+    fromServerPeon.markSegmentToDrop(segment);
+    segmentsMovingInTier.addSegment(segment.getId(), fromServer.getServer().getHost());
+
+    final LoadQueuePeon toServerPeon = toServer.getPeon();
+    final String toServerName = toServer.getServer().getName();
+    try {
+      toServerPeon.loadSegment(
+          segment,
+          SegmentAction.MOVE_TO,
+          success -> {
+            // Drop segment only if:
+            // (1) segment load was successful on toServer
+            // AND (2) segment not already queued for drop on fromServer
+            // AND (3a) loading is http-based
+            //     OR (3b) inventory shows segment loaded on toServer
+
+            // Do not check the inventory with http loading as the HTTP
+            // response is enough to determine load success or failure
+            if (success
```

Comment:
If it wasn't successful, do we need to update the state that was set from the call to `fromServerPeon.markSegmentToDrop`?

Reply:
Yes, it happens through the `moveFinishCallback.execute()` in the `else` block.